cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala

```diff
@@ -0,0 +1,104 @@
+        val copyCommand = Seq("bash", "-c", generateDownloadScript(gcsUrl, Option(saJsonPath)))
+        Process(copyCommand, None, extraEnv.toSeq: _*)
+      case None =>
+        // No SA returned from Martha. gsutil will use the application default credentials.
+        val copyCommand = Seq("bash", "-c", generateDownloadScript(gcsUrl, None))
+        Process(copyCommand)
+    }
+
+    // run the multiple bash script to download file and log stream sent to stdout and stderr using ProcessLogger
+    val returnCode = copyProcess ! ProcessLogger(logger.underlying.info, logger.underlying.error)
+
+    IO(ExitCode(returnCode))
+  }
+
+  /**
+    * Bash to download the GCS file using `gsutil`.
+    */
+  def generateDownloadScript(gcsUrl: String, saJsonPathOption: Option[Path]): String = {
+
+    def gcsCopyCommand(flag: String = ""): String = s"gsutil $flag cp $gcsUrl $downloadLoc"
+
+    def setServiceAccount(): String = {
+      saJsonPathOption match {
+        case Some(saJsonPath) =>
+          s"""# Set gsutil to use the service account returned from Martha
+             |gcloud auth activate-service-account --key-file=$saJsonPath > gcloud_output.txt 2>&1
+             |RC_GCLOUD=$$?
+             |if [ "$$RC_GCLOUD" != "0" ]; then
+             |  echo "Failed to activate service account returned from Martha. File won't be downloaded. Error: $$(cat gcloud_output.txt)" >&2
+             |  exit "$$RC_GCLOUD"
+             |else
+             |  echo "Successfully activated service account; Will continue with download. $$(cat gcloud_output.txt)"
+             |fi
+             |""".stripMargin
+        case None => ""
+      }
+    }
+
+    def recoverWithRequesterPays(): String = {
+      requesterPaysProjectIdOption match {
+        case Some(userProject) =>
+          s"""if [ "$$RC_GSUTIL" != "0" ]; then
+             |  # Check if error is requester pays. If yes, retry gsutil copy using project flag
+             |  if grep -q '$RequesterPaysErrorMsg' gsutil_output.txt; then
+             |    echo "Received 'Bucket is requester pays' error. Attempting again using Requester Pays billing project"
+             |    ${gcsCopyCommand(s"-u $userProject")} > gsutil_output.txt 2>&1
+             |    RC_GSUTIL=$$?
+             |  fi
+             |fi""".stripMargin
+        case None => ""
+      }
+    }
+
+    // bash to download the GCS file using gsutil
+    s"""set -euo pipefail
+       |set +e
+       |
+       |${setServiceAccount()}
+       |
+       |# Run gsutil copy without using project flag
+       |${gcsCopyCommand()} > gsutil_output.txt 2>&1
+       |RC_GSUTIL=$$?
+       |
+       |${recoverWithRequesterPays()}
+       |
+       |if [ "$$RC_GSUTIL" != "0" ]; then
+       |  echo "Failed to download the file. Error: $$(cat gsutil_output.txt)" >&2
+       |  exit "$$RC_GSUTIL"
+       |else
+       |  echo "Download complete!"
+       |  exit 0
+       |fi""".stripMargin
```

Comment:
2 of 2: Any opinions on this block terminating with a newline? (the change would affect the test expectations too.)
```scala
|fi
|""".stripMargin
```

This one terminates the actual full bash script. When I'm not trying to code golf I generally [prefer newlines](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file#answer-18789).

Reply:
⛳